lib/ruby-mpd/plugins/command_list.rb

```diff
@@ -0,0 +1,177 @@
+module MPD::Plugins
+
+  # Batch send multiple commands at once for speed.
+  module CommandList
+    # Send multiple commands at once.
+    #
+    # By default, any response from the server is ignored (for speed).
+    # To get results, pass +response_type+ as one of the following options:
+    #
+    # * +:hash+   — a single hash of return values; multiple return values for the same key are grouped in an array
+    # * +:values+ — an array of individual values
+    # * +:hashes+ — an array of hashes (where value boundaries are guessed based on the first result)
+    # * +:songs+  — an array of Song instances from the results
+    # * +:playlists+ - an array of Playlist instances from the results
+    #
+    # Note that each supported command has no return value inside the block.
+    # Instead, the block itself returns the array of results.
+    #
+    # @param [Symbol] response_type the type of responses desired.
+    # @return [nil] default behavior.
+    # @return [Array] if +response_type+ is +:values+, +:hashes+, +:songs+, or +:playlists+.
+    # @return [Hash] if +response_type+ is +:hash+.
+    #
+    # @example Simple batched control commands
+    #   @mpd.command_list do
+    #     stop
+    #     shuffle
+    #     save "shuffled"
+    #   end
+    #
+    # @example Adding songs to the queue, ignoring the response
+    #   @mpd.command_list do
+    #     my_songs.each do |song|
+    #       add(song)
+    #     end
+    #   end
+    #
+    # @example Adding songs to the queue and getting the song ids
+    #   ids = @mpd.command_list(:values){ my_songs.each{ |song| addid(song) } }
+    #   #=> [799,800,801,802,803]
+    #
+    #   ids = @mpd.command_list(:hashes){ my_songs.each{ |song| addid(song) } }
+    #   #=> [ {:id=>809}, {:id=>810}, {:id=>811}, {:id=>812}, {:id=>813} ]
+    #
+    #   ids = @mpd.command_list(:hash){   my_songs.each{ |song| addid(song) } }
+    #   #=> { :id=>[804,805,806,807,808] }
+    #
+    # @example Finding songs matching various genres
+    #   songs = @mpd.command_list(:songs) do
+    #     where genre:'Alternative Rock'
+    #     where genre:'Alt. Rock'
+    #     where genre:'alt-rock'
+    #   end
+    #
+    # @see CommandList::Commands CommandList::Commands for a list of supported commands.
+    def command_list(response_type=nil,&commands)
+      @mutex.synchronize do
+        begin
+          socket.puts "command_list_begin"
+          CommandList::Commands.new(self).instance_eval(&commands)
```

Comment:
In the case explicitly where `response_type==nil` I can imagine doing what you suggest: starting another thread, making a new connection (basically a new MPD instance matching the same settings) and sending off the commands.

Given that we do not know if this would provide any real benefit (both because of no real concurrency, and the overhead of establishing a new connection), and it's a moderate amount of additional code and work, I'd suggest that it's a theoretically good idea that should be investigated in the future separate from this pull request.